lib/archethic/transaction_chain/transaction/data/recipient.ex

```diff
@@ -0,0 +1,110 @@
+defmodule Archethic.TransactionChain.TransactionData.Recipient do
+  @moduledoc """
+  Represents a call to a Smart Contract
+
+  Action & Args are nil for a :transaction trigger and are filled for a {:transaction, action, args} trigger
+  """
+  alias Archethic.Crypto
+  alias Archethic.Utils
+
+  defstruct [:address, :action, :args]
+
+  @unnamed_action 0
+  @named_action 1
+
+  @type t :: %__MODULE__{
+          address: Crypto.prepended_hash(),
+          action: String.t(),
+          args: list(any())
+        }
+
+  @doc """
+  Return wether this is a named action call or not
+  """
+  @spec is_named_action?(t()) :: boolean()
+  def is_named_action?(%__MODULE__{action: nil, args: nil}), do: false
+  def is_named_action?(%__MODULE__{}), do: true
+
+  @doc """
+  Serialize a recipient
+  """
+  @spec serialize(t(), pos_integer()) :: binary()
+  def serialize(%__MODULE__{address: address}, 1) do
+    <<address::binary>>
+  end
+
+  def serialize(%__MODULE__{address: address, action: nil, args: nil}, 2) do
+    <<@unnamed_action::8, address::binary>>
+  end
+
+  def serialize(%__MODULE__{address: address, action: action, args: args}, 2) do
+    # 255 chars should be enough
+    action_bytes = byte_size(action)
+    true = 255 >= action_bytes
```

Comment:
This should be checked before.
In `TransactionPayload` or even when parsing the contract.